client/network/src/protocol/generic_proto/upgrade/notifications.rs

```diff
@@ -173,6 +173,48 @@ where TSubstream: AsyncRead + AsyncWrite,
 
 		self.handshake = NotificationsInSubstreamHandshake::PendingSend(message.into());
 	}
+
+	/// Equivalent to `Stream::poll_next`, except that it only drives the handshake and is
+	/// guaranteed to not generate any notification.
+	pub fn poll_process(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<Infallible, io::Error>> {
+		let mut this = self.project();
+
+		loop {
+			match mem::replace(this.handshake, NotificationsInSubstreamHandshake::Sent) {
+				NotificationsInSubstreamHandshake::PendingSend(msg) =>
```

Comment:
Instead of duplicating the handshake sending logic here, how about moving all of the polling logic into a new function (e.g. `poll_inner`) which takes a parameter `handshake_only`. `poll_process` would invoke `poll_inner` with `true`, `Stream::poll` would do so with `false`. What do you think?

Reply:
`poll_next` and `poll_process` have different return types, with the return type of `poll_process` being more restrictive.
If I merge them into `poll_inner`, I then have to add some `unreachable!()`s which I'm not sure is an improvement.